Kernel/IPC: Use Ports and Sessions as the fundamental building block of Inter Process Communication.

Subv opened the pull request:
This is an overhaul of #1911 and a continuation of #1842

This PR is part of an ongoing effort to implement multiprocess support in Citra.

All handles obtained via srv::GetServiceHandle or svcConnectToPort are references to ClientSessions.
Service modules will wait on the counterpart of those ClientSessions (Called ServerSessions) using svcReplyAndReceive or svcWaitSynchronization[1|N], and will be awoken when a SyncRequest is performed.

All IPC parameter translation for LLE modules will happen inside the implementation of svcReplyAndReceive, this is not implemented in this PR.

All IPC parameter translation for HLE modules should happen in the SessionRequestHandler class, this is not implemented in this PR.

HLE Interfaces are no longer kernel objects, and now derive from SessionRequestHandler, they are bound to a Session pair when a connection request is initiated by the emulated application (See svcConnectToPort and srv:GetServiceHandle).

There is currently a very important unimplemented feature related to #2125. SyncRequests should put the calling thread to sleep until the server counterpart of the session handles and answers the request (by means of svcReplyAndReceive), but this is all instantaneous in our current HLE modules.
It is also speculated that the real kernel preempts the threads running in the AppCore and reschedules to another thread if said thread was waiting on a SyncRequest and the request was completed. (This speculation comes from the aforementioned #2125 and reverse engineering of the svcReplyAndReceive handler in the ARM11 kernel)

# Notable changes
* HLE Interfaces are no longer kernel objects
* HLE Interfaces are now attached to ServerSessions, and are called during ServerSession::HandleSyncRequest
* HLE Interfaces now all inherit from SessionRequestHandler.
* GetServiceHandle and ConnectToPort now properly return ClientSession handles to the emulated application.

# Future changes
- [ ] Support session disconnection.
- [ ] Implement svcAcceptSession, it would only be used by the LLE services.
- [ ] Implement svcReplyAndReceive, it would only be used by the LLE services. This call works pretty much like WaitSynchronizationN.
- [ ] Rewrite how WaitSynchronizationN works. The current code is confusing and may lead to errors. (See #2260)
- [ ] Implement command buffer translation inside svcReplyAndReceive.
- [ ] Copy the command buffer to a local buffer before calling HandleSyncRequest in SessionRequestHandler.

Subv commented:
cc @Lectem

Subv commented on src/core/hle/kernel/server_session.h:
```diff
@@ -0,0 +1,77 @@
+ * the request, the response is marshalled back to the caller's TLS buffer and control is
+ * transferred back to it.
+ */
+class ServerSession : public WaitObject {
```
This class should contain a reference to the last thread that initiated a SyncRequest for the associated ClientSession so we can access it's TLS.

Subv commented on src/core/hle/service/service.cpp:
```diff
@@ -61,7 +64,9 @@ static std::string MakeFunctionString(const char* name, const char* port_name,
 
-ResultVal<bool> Interface::SyncRequest() {
+ResultCode Interface::HandleSyncRequest(Kernel::SharedPtr<Kernel::ServerSession> server_session) {
+    // TODO(Subv): Make use of the server_session in the HLE service handlers to distinguish which session triggered each command.
```
How should we approach this TODO?

Lectem replied:
Is it important to know what session triggered each command ? I thought that it was handled by creating a new thread and storing session related information in the TLS (I might be entirely off the mark for this one)

Subv replied:
Multiple applications could be doing requests to the same service at the same time, we need to know which thread to awake and who we're responding to.
Real services are allowed to do whatever they want, most do spawn a new thread for each request, however others just handle them all in the same thread.

yuriks replied:
You can just stick the session data in a map (keyed by the pointer) to maintain state.

Subv commented on src/core/hle/service/fs/archive.cpp:
```diff
@@ -236,10 +237,10 @@ ResultVal<bool> Directory::SyncRequest() {
         LOG_ERROR(Service_FS, "Unknown command=0x%08X!", cmd);
         ResultCode error = UnimplementedFunction(ErrorModule::FS);
         cmd_buff[1] = error.raw; // TODO(Link Mauve): use the correct error code for that.
-        return MakeResult<bool>(false);
```
Is this a  bug? in the File::SyncRequest default case we return the error instead of RESULT_SUCCESS.

wwylele replied:
Not really sure what bug you are referring to... does real 3DS return RESULT_SUCCESS on unknown command?

Subv replied:
No idea, but we do different things for unknown File commands and unknown Directory commands.

Subv commented on src/core/hle/service/fs/archive.cpp:
```diff
@@ -172,7 +173,8 @@ ResultVal<bool> File::SyncRequest() {
 
     case FileCommand::OpenLinkFile: {
         LOG_WARNING(Service_FS, "(STUBBED) File command OpenLinkFile %s", GetName().c_str());
```
I don't think this is a stub anymore, after this change

yuriks replied:
Are you removing it?

Subv replied:
I'd rather wait until we actually reverse the function to see what it does.

Subv commented on src/core/hle/service/service.cpp:
```diff
@@ -75,14 +80,14 @@ ResultVal<bool> Interface::SyncRequest() {
     itr->second.func(this);
 
-    return MakeResult<bool>(false); // TODO: Implement return from actual function
+    return RESULT_SUCCESS; // TODO: Implement return from actual function, it should fail if the parameter translation fails
```
The second part of this TODO should be handled by the Translate() call.

Subv commented on src/core/hle/service/service.h:
```diff
@@ -9,29 +9,205 @@
+class ServerSession;
+
+// TODO(Subv): Move these declarations out of here
+static const int kCommandHeaderOffset = 0x80; ///< Offset into command buffer of header
```
Where would be a good place to put these?

Lectem replied:
I think it should be an internal of the IPC command helper

yuriks replied:
This will be used by the kernel translation code too, so it should stay in a more generic file.

Subv commented on src/core/hle/service/service.h:
```diff
@@ -9,29 +9,205 @@
+}
+
+// TODO(Subv): Move this namespace out of here
+namespace IPC {
```
Where would be a good place to put these?

Lectem replied:
if #2027 gets merged, it will already be moved to ipc.h

yuriks replied:
Are you keeping these in this file after all?

Subv replied:
No, I'll move them to an ipc.h header

Subv commented on src/core/hle/service/service.h:
```diff
@@ -9,29 +9,205 @@
+     * Dispatches and handles a sync request from the emulated application.
+     * @param server_session The ServerSession that was triggered for this sync request,
+     * it should be used to differentiate which client (As in ClientSession) we're answering to.
+     * TODO(Subv): Make a HandleSyncRequestParent function that is called from the outside and does { ReturnIfError(Translate()); HandleSyncRequest(); }
```
Thoughts on this? @Lectem @yuriks

Lectem replied:
> it should be used to differentiate which client (As in ClientSession) we're answering to.

I think each server session should map to a TLS, and the service is responsible for choosing which one (by creating a thread) if it is needed.

Subv replied:
See my previous response.

yuriks replied:
> TODO(Subv): Make a HandleSyncRequestParent function that is called from the outside and does { ReturnIfError(Translate()); HandleSyncRequest(); }

Nay. This class only applies to HLE stuff, and at least for the moment we won't be doing any translation on them. Real calls won't even go through here, so you'll need some other translation call else where. I think this should be handled on ServerSession::HandleSyncRequest or a similar place.

Subv replied:
The translation (both client-server and server-client) has to happen during svcReplyAndReceive for non-HLE stuff.

I think the translation codepaths for HLE and non-HLE are different enough to warrant separating them and calling the HLE version from here

Subv commented on src/core/hle/service/service.h:
```diff
@@ -9,29 +9,205 @@
 static const int kMaxPortSize = 8; ///< Maximum size of a port name (8 characters)
+static const u32 DefaultMaxSessions = 10; ///< Arbitrary default number of maximum connections to an HLE port
+
+/// TODO(Subv): Write documentation for this class
```
I'm not sure what to write here.

yuriks replied:
> Interface implemented by HLE Session handlers. This can be provided to a Session in order to hook into several relevant events (such as a new connection or a SyncRequest) so they can be implemented in the emulator.

...or something like that?

Subv commented on src/core/hle/service/service.h:
```diff
@@ -9,29 +9,205 @@
+    virtual ResultCode HandleSyncRequest(Kernel::SharedPtr<Kernel::ServerSession> server_session) = 0;
+};
 
 /// Interface to a CTROS service
```
This comment is not descriptive at all, what would be a better docstring for this class?

yuriks replied:
> Framework for implementing HLE service handlers which dispatch incoming SyncRequests based on a table mapping header ids to handler functions. *etc.* *etc.*

Subv commented on src/core/hle/service/srv.cpp:
```diff
@@ -81,7 +85,23 @@ static void GetServiceHandle(Service::Interface* self) {
+        auto client_session = std::get<Kernel::SharedPtr<Kernel::ClientSession>>(sessions);
+        auto server_session = std::get<Kernel::SharedPtr<Kernel::ServerSession>>(sessions);
+
+        // TODO(Subv): Wait the current thread until the ServerPort calls AcceptSession.
```
After further inspecting of the ARM11 kernel code, it appears this call should not put the thread to sleep, more investigation is needed.

Subv commented:
It has been verified with a hardware test that GetServiceHandle and ConnectToPort do not put the calling thread to sleep until AcceptSession has been called.

lioncash commented on src/core/hle/kernel/client_session.cpp:
```diff
@@ -0,0 +1,28 @@
+
+namespace Kernel {
+
+ClientSession::ClientSession() {}
```
You can default these as:

```cpp
ClientSession::ClientSession() = default;
...
```

lioncash commented on src/core/hle/kernel/server_session.cpp:
```diff
@@ -0,0 +1,56 @@
+
+namespace Kernel {
+
+ServerSession::ServerSession() {}
```
ditto on defaulting.

Lectem commented on src/core/hle/service/fs/archive.cpp:
```diff
@@ -92,7 +93,7 @@ File::File(std::unique_ptr<FileSys::FileBackend>&& backend, const FileSys::Path&
 
-ResultVal<bool> File::SyncRequest() {
+ResultCode File::HandleSyncRequest(Kernel::SharedPtr<Kernel::ServerSession> server_session) {
     u32* cmd_buff = Kernel::GetCommandBuffer();
```
We should use the IPC helper from #2027 to wrap the command buffer, and make it accessible from the ServerSession, or (better I think?) simply replace the pointer to the session by the helper

Subv commented on src/core/hle/kernel/client_port.cpp:
```diff
@@ -6,10 +6,17 @@
 ClientPort::~ClientPort() {}
 
+void ClientPort::AddWaitingSession(SharedPtr<ServerSession> server_session) {
+    server_port->pending_sessions.push_back(server_session);
```
Should we check if the ServerSession has an associated hle_handler and not add it to the pending list? Or just add it anyways? (Will we ever use this list in HLE services?)

yuriks commented on src/core/hle/service/service.h:
```diff
@@ -9,29 +9,205 @@
+     * it should be used to differentiate which client (As in ClientSession) we're answering to.
+     * TODO(Subv): Make a HandleSyncRequestParent function that is called from the outside and does { ReturnIfError(Translate()); HandleSyncRequest(); }
+     * The Translate() function would copy the command buffer from the ServerSession thread's TLS into a temporary buffer, and pass it to HandleSyncRequest.
+     * TODO(Subv): HandleSyncRequest's return type should be void.
```
Is it too difficult to just do this right now? It's mostly just deleting lines from a few places you're already modifying in this PR anyway.

yuriks commented on src/core/hle/service/service.h:
```diff
@@ -9,29 +9,205 @@
+     * The Translate() function would copy the command buffer from the ServerSession thread's TLS into a temporary buffer, and pass it to HandleSyncRequest.
+     * TODO(Subv): HandleSyncRequest's return type should be void.
+     */
+    virtual ResultCode HandleSyncRequest(Kernel::SharedPtr<Kernel::ServerSession> server_session) = 0;
```
I recommend making a struct (`RequestParams` or so) to hold the parameters. That way we can expand it in the future without having to go back and fix every single handler to handle the new parameters. (This can in the future hold stuff such as the translated cmdbuf, the originating thread, etc.)

yuriks commented on src/core/hle/service/service.h:
```diff
@@ -9,29 +9,205 @@
+     * It should be overwritten by each service implementation for more fine-grained control.
+     * @returns The maximum number of connections allowed.
+     */
+    virtual u32 GetMaxSessions() const { return DefaultMaxSessions; }
```
No reason to make this a virtual function, just make it a constructor parameter so subclasses can set it.

yuriks commented on src/core/hle/service/srv.cpp:
```diff
@@ -81,7 +85,24 @@ static void GetServiceHandle(Service::Interface* self) {
+        // AcceptSession before returning from this call.
+
+        // Add the server session to the port's queue
+        client_port->AddWaitingSession(server_session);
```
Lot's of code duplicated between here and the ConnectToPort svc. See comment in ClientPort::AddWaitingSession.

Subv commented on src/core/hle/kernel/server_session.cpp:
```diff
@@ -0,0 +1,56 @@
+
+ResultCode ServerSession::HandleSyncRequest() {
+    // The ServerSession received a sync request, this means that there's new data available
+    // from its ClientSession, so wake up any threads that may be waiting on a svcReplyAndReceive or similar.
```
It's just a description of the function.
We received data, so wake the waiting threads

yuriks replied:
Hm, yeah, not sure what I was thinking before, since you do wake up the threads. nvm

Subv commented on src/core/hle/kernel/server_session.cpp:
```diff
@@ -0,0 +1,56 @@
+
+    // If this ServerSession does not have an HLE implementation, just wake up the threads waiting on it.
+    signaled = true;
+    WakeupAllWaitingThreads();
```
Yep

Subv commented on src/core/hle/kernel/server_session.cpp:
```diff
@@ -0,0 +1,56 @@
+    signaled = true;
+    WakeupAllWaitingThreads();
+    return RESULT_SUCCESS;
+}
```
This function doesnt have any missing functionality that i can think of right now, what do you mean?

yuriks replied:
nvm

Subv commented:
Addressed some comments.
The Session disconnection logic is still todo, but I think that can be done in a follow up PR so as to not make this one bigger than it already is.

yuriks commented on src/core/hle/kernel/server_port.cpp:
```diff
@@ -24,12 +24,13 @@ void ServerPort::Acquire() {
     SharedPtr<ClientPort> client_port(new ClientPort);
 
     server_port->name = name + "_Server";
+    server_port->hle_handler = hle_handler;
```
~ std::move ~

yuriks commented on src/core/hle/kernel/server_session.cpp:
```diff
@@ -0,0 +1,56 @@
+
+    // If this ServerSession has an associated HLE handler, forward the request to it.
+    if (hle_handler != nullptr)
+        return hle_handler->HandleSyncRequest(SharedPtr<ServerSession>(this));
```
Is there any reason why you shouldn't also signal this session if it's HLE?

Subv replied:
Nothing would be waiting on the ServerSession if it's HLE, it wouldn't even be exposed to the running application, there's no point in doing that (I think?)

yuriks replied:
But is there any reason not to do it? That's one less difference in case we ever move HLE to be more process-based instead of callback based.

Subv replied:
Done~~

yuriks commented on src/core/hle/kernel/server_session.cpp:
```diff
@@ -0,0 +1,56 @@
+    return RESULT_SUCCESS;
+}
+
+std::tuple<SharedPtr<ServerSession>, SharedPtr<ClientSession>> ServerSession::CreateSessionPair(const std::string& name, std::shared_ptr<Service::SessionRequestHandler> hle_handler) {
```
Overlong line

yuriks commented on src/core/hle/kernel/server_session.cpp:
```diff
@@ -0,0 +1,56 @@
+}
+
+std::tuple<SharedPtr<ServerSession>, SharedPtr<ClientSession>> ServerSession::CreateSessionPair(const std::string& name, std::shared_ptr<Service::SessionRequestHandler> hle_handler) {
+    auto server_session = ServerSession::Create(name + "_Server", hle_handler).MoveFrom();
```
move hle_handler

yuriks commented on src/core/hle/service/fs/archive.h:
```diff
@@ -41,33 +41,37 @@ enum class MediaType : u32 { NAND = 0, SDMC = 1 };
 typedef u64 ArchiveHandle;
 
-class File : public Kernel::Session {
+class File : public SessionRequestHandler, public std::enable_shared_from_this<File> {
```
Mark as final

yuriks commented on src/core/hle/service/fs/archive.h:
```diff
@@ -41,33 +41,37 @@ enum class MediaType : u32 { NAND = 0, SDMC = 1 };
 };
 
-class Directory : public Kernel::Session {
+class Directory : public SessionRequestHandler {
```
Mark as final

yuriks commented on src/core/hle/service/srv.cpp:
```diff
@@ -81,7 +85,20 @@ static void GetServiceHandle(Service::Interface* self) {
+        auto client_port = it->second;
+
+        // Note: Threads do not wait for the server endpoint to call
+        // AcceptSession before returning from this call.
```
Noticed that this comment is duplicated in two places. Maybe move it to the Connect function.

yuriks commented on src/core/hle/service/srv.cpp:
```diff
@@ -81,7 +85,20 @@ static void GetServiceHandle(Service::Interface* self) {
+        // Note: Threads do not wait for the server endpoint to call
+        // AcceptSession before returning from this call.
+
+        // Connect to the port and retrieve the client endpoint of the connection Session.
```
Seems like a kinda superfluous comment, it's just re-stating the code line.

yuriks commented on src/core/hle/svc.cpp:
```diff
@@ -222,20 +224,31 @@ static ResultCode ConnectToPort(Handle* out_handle, const char* port_name) {
     return RESULT_SUCCESS;
 }
 
 /// Synchronize to an OS service
```
"Makes a blocking IPC call to an OS service."

yuriks commented on src/core/hle/service/service.cpp:
```diff
@@ -61,7 +64,29 @@ static std::string MakeFunctionString(const char* name, const char* port_name,
+}
+
+void Interface::HandleSyncRequestImpl(Kernel::SharedPtr<Kernel::ServerSession> server_session) {
+    // TODO(Subv): Make use of the server_session in the HLE service handlers to distinguish which session triggered each command.
```
Wouldn't this be more of a per-service level concern? Each one might want to do different things, so this doesn't seem like an appropriate place for this todo.

Subv commented on src/core/hle/service/service.h:
```diff
@@ -9,30 +9,235 @@
+     * it should be used to differentiate which client (As in ClientSession) we're answering to.
+     * @returns ResultCode the result code of the translate operation.
+     */
+    ResultCode HandleSyncRequest(Kernel::SharedPtr<Kernel::ServerSession> server_session);
```
I don't agree, where would we put it? IMO this class should take care of the translation for all HLE IPC requests, and just hand a translated buffer to the implementation.

yuriks commented on src/core/hle/service/service.cpp:
```diff
@@ -84,6 +84,9 @@ ResultCode SessionRequestHandler::TranslateRequest(Kernel::SharedPtr<Kernel::Ser
 }
 
+Interface::~Interface() = default;
+Interface::Interface(u32 max_sessions) : max_sessions(max_sessions) { }
```
Constructor goes before the destructor

Subv commented:
@yuriks Do you have anything against the ClientConnected / ClientDisconnected callbacks? I thought you were in favor of this idea.

Subv commented:
A @yuriks LGTM means this thing is out of RFC state \o/

Subv commented:
Formatting errors have been fixed, can this be merged?
